quinn/src/endpoint.rs

```diff
@@ -349,22 +350,28 @@ where
                 }
             }
             if self.outgoing.is_empty() {
+                self.send_limiter.finish_cycle();
                 return Ok(false);
             }
+
+            if !self.send_limiter.allow_work() {
+                self.send_limiter.finish_cycle();
+                return Ok(true);
+            }
+
             match self.socket.poll_send(cx, self.outgoing.as_slices().0) {
                 Poll::Ready(Ok(n)) => {
                     self.outgoing.drain(..n);
                     // We count transmits instead of `poll_send` calls since the cost
                     // of a `sendmmsg` still linearily increases with number of packets.
-                    transmits += n;
-                    if transmits >= IO_LOOP_BOUND {
-                        return Ok(true);
-                    }
+                    self.send_limiter.record_work(n);
                 }
                 Poll::Pending => {
+                    self.send_limiter.finish_cycle();
                     return Ok(false);
                 }
                 Poll::Ready(Err(e)) => {
+                    self.send_limiter.finish_cycle();
```

Comment:
All these separate exits make me start leaning more towards @djc's favored `Drop` guard approach for bounding cycle. Might be easy to miss one in a refactor.

Reply:
Right - I didn't like the duplication too much either. Now I improved it a bit by breaking from the loop and having a common return place. Doesn't protect against someone acccidentally adding a `return` later on, but provides some decent improvement for the amount of changes.